service/src/main/java/uk/gov/hmcts/reform/fpl/service/legalcounsel/ManageLegalCounselService.java

```diff
@@ -0,0 +1,169 @@
+package uk.gov.hmcts.reform.fpl.service.legalcounsel;
+
+import com.google.common.base.Strings;
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.fpl.enums.SolicitorRole;
+import uk.gov.hmcts.reform.fpl.events.LegalCounsellorAdded;
+import uk.gov.hmcts.reform.fpl.events.LegalCounsellorRemoved;
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.model.Child;
+import uk.gov.hmcts.reform.fpl.model.LegalCounsellor;
+import uk.gov.hmcts.reform.fpl.model.Respondent;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.fpl.model.event.ManageLegalCounselEventData;
+import uk.gov.hmcts.reform.fpl.model.interfaces.WithSolicitor;
+import uk.gov.hmcts.reform.fpl.service.CaseConverter;
+import uk.gov.hmcts.reform.fpl.service.CaseRoleLookupService;
+import uk.gov.hmcts.reform.fpl.service.EventService;
+import uk.gov.hmcts.reform.fpl.service.OrganisationService;
+import uk.gov.hmcts.reform.rd.model.Organisation;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyList;
+import static java.util.function.Predicate.not;
+import static uk.gov.hmcts.reform.fpl.enums.SolicitorRole.Representing.CHILD;
+import static uk.gov.hmcts.reform.fpl.enums.SolicitorRole.Representing.RESPONDENT;
+
+@RequiredArgsConstructor
+@Component
+public class ManageLegalCounselService {
+
+    private static final List<SolicitorRole.Representing> RELEVANT_REPRESENTED_PARTY_TYPES = asList(RESPONDENT, CHILD);
+
+    private static final String UNREGISTERED_USER_ERROR_MESSAGE_TEMPLATE = "Unable to grant access "
+        + "[%s is not a Registered User] - Email address for Legal representative is not registered on the system. "
+        + "They can register at https://manage-org.platform.hmcts.net/register-org/register";
+
+    private final CaseConverter caseConverter;
+    private final CaseRoleLookupService caseRoleLookupService;
+    private final OrganisationService organisationService;
+    private final EventService eventPublisher;
+
+    public List<Element<LegalCounsellor>> retrieveLegalCounselForLoggedInSolicitor(CaseDetails caseDetails) {
```

Comment:
Why do all of these methods accept case details?

Reply:
It was a lot cleaner on the tests and easier when you need both the CaseData and the Map representation of it.
Happy to talk about it if you still think it should take CaseDetails as a parameter.

Reply:
I've applied this change to most methods, apart from the one we talked about.